rustc_codegen_llvm: Feature Conversion Tidying

a4lg opened the pull request:
The author thinks we can improve `to_llvm_features`, a function to convert a Rust target feature name into an LLVM feature (or nothing, to ignore features unsupported by LLVM) for better maintainability.

1.  We can simplify some clauses and some expressions.
2.  There are some readability issues.

This PR attempts to resolve some of them by tidying many cases.

<!-- homu-ignore:start -->

---------

**OUTDATED CONTENTS FOLLOW**

# Commit Groups

There are three kinds of changes in this PR:

1.  Simplify certain places (Commit 1–3)  
    I think merging them should be (simply) okay.
2.  Merge match clauses (Commit 4)  
    While many would accept, some changes might break the intent of the original author(s).
3.  Reorder conversion cases (Commit 5)  
    I'll describe this later.

**To reviewers:** I would like to hear which commits you can accept (if not all).

# Reordering (Commit 5)

While match clauses of the current implementation seem too random, I think there are multiple choices to take.

1.  Just sort by the architecture.  
    It's simpler and we don't need to separate `avx10.1` and `avx10.2` into two different subsections.  However, we will have similar match clauses for pretty much the same purpose (to filter out features unsupported by LLVM) in multiple places.
2.  First sort by the conversion category and then the architecture (as in this PR).  
    If we go with it, there are some cases we will have to separate similar target features into multiple categories (like `avx10.1` with simple renaming and `avx10.2` with LLVM version filter + renaming (considered complex) in `x86`).  Instead, we will have a clear view of what's going to happen on the specific target feature.

I chose (2) but (1) is definitely a good option, too.
<!-- homu-ignore:end -->

workingjubilee commented on compiler/rustc_codegen_llvm/src/llvm_util.rs:
```diff
@@ -288,43 +295,32 @@ pub(crate) fn to_llvm_features<'a>(sess: &Session, s: &'a str) -> Option<LLVMFea
         )),
+        ("x86", "avx10.2") => (get_version().0 >= 20).then(|| LLVMFeature::new("avx10.2-512")),
+        /*
+            By default, use the Rust's target feature name as-is.
```
"the Rust's"?

a4lg replied:
Simple mistake.
Since I abandon the approach 2 on the commit 5 (based on your another feedback), this comment will be removed.

workingjubilee commented on compiler/rustc_codegen_llvm/src/llvm_util.rs:
```diff
@@ -288,43 +295,32 @@ pub(crate) fn to_llvm_features<'a>(sess: &Session, s: &'a str) -> Option<LLVMFea
+            Perform relatively complex feature conversion.
+        */
+        // Before LLVM 20 those two features were packaged together as b16b16
+        ("aarch64", "sve-b16b16" | "sme-b16b16") if get_version().0 < 20 => {
```
Most of these instances use the 3-tuple. I would prefer normalizing in the other direction, despite the technical redundancy.

a4lg replied:
For now, I will abandon the commit 4.

workingjubilee commented on compiler/rustc_codegen_llvm/src/llvm_util.rs:
```diff
@@ -288,43 +295,32 @@ pub(crate) fn to_llvm_features<'a>(sess: &Session, s: &'a str) -> Option<LLVMFea
+            "x86",
+            "amx-avx512" | "amx-fp8" | "amx-movrs" | "amx-tf32" | "amx-transpose" | "movrs",
+        ) if get_version().0 < 20 => None,
+        /*
+            Perform relatively complex feature conversion.
+        */
```
This doesn't make sense to me because "simple" and "complex" are purely subjective here, and this redistributes the match arms for features for a single architecture across the match instead of putting them together. This is not a coherent organization strategy that makes sense with how patches usually land. Typically, architecturally focused patches affect one architecture at a time, so they should all be closer together.

workingjubilee reviewed (commented):
Why do you switch from "The author" 3rd-person-self-reference to "I"?

Is this anything but just churn?

Did you ask a machine to output the PR description text according to a prompt and then paste it in?

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/146615/a454fccb02df9d361f1201b747c01257f58a8b37..65387112e437714982a24c5a94b0902296be32d1)*

workingjubilee commented:
As someone who has made several PRs and landed them, most of them on an arch-by-arch basis, you should already know this organization strategy doesn't make sense.

workingjubilee commented:
r? @workingjubilee

a4lg commented:
@workingjubilee What about commits 1 through 3?  If they are working for you, I will separate the PR into (a) commit 1–3 and (b) heavily modified commit 5 (sorted only by the architecture).

workingjubilee commented:
6361abcd58249137403e6460996ed537f6cb61c0 looks fine.

The merged matches are probably fine.

Otherwise my remark about comparing against the 3-tuple instead of `.0` applies: the `.0` doesn't make it clear that it's comparing against the major of a `(major, minor, patch)` tuple.

a4lg commented:
> As someone who has made several PRs and landed them, most of them on an arch-by-arch basis, you should already know this organization strategy doesn't make sense.

Understood, taking another approach (sort only by the architecture) in the next submission.

a4lg committed:
rustc_codegen_llvm: Reorder conversion cases

For maintainability, this commit reorders target feature conversion
cases by the architecture.

a4lg committed:
rustc_codegen_llvm: Simplify `arch` conversion

This commit simplifies construction of `arch` from `sess.target.arch`.
It also preserves a reference to `sess.target.arch` as `raw_arch`
to make this function future proof.

a4lg committed:
rustc_codegen_llvm: Name major version of LLVM

It makes LLVM version comparison clearer.

workingjubilee commented on compiler/rustc_codegen_llvm/src/llvm_util.rs:
```diff
@@ -217,27 +217,16 @@ impl<'a> IntoIterator for LLVMFeature<'a> {
+        "powerpc64" => "powerpc",
+        _ => raw_arch,
     };
+    let (major, _, _) = get_version();
```
yeah that's also good.

workingjubilee commented:
Thanks.

@bors r+ rollup
